[SYCL][Bindless][Exp] Add Support For Unsampled Image Arrays

isaacault opened the pull request:
- Creation / destruction of unsampled image arrays
- Fetching / writing of unsampled image arrays
- `sycl::ext::oneapi::experimental::image_type::array` enum value added
- `sycl::ext::oneapi::experimental::image_descriptor::array_size` member added
- `sycl::ext::oneapi::experimental::image_descriptor::verify()` member function added

Correlated UR PR: [[Bindless][Exp] Add Support For Image Arrays #1274](https://github.com/oneapi-src/unified-runtime/pull/1274)

cppchedy commented on sycl/test-e2e/bindless_images/bindless_helpers.hpp:
```diff
@@ -0,0 +1,59 @@
+
+template <typename DType, int NChannels,
+          typename = std::enable_if_t<NChannels == 1>>
+static DType add_kernel(const DType in_0, const DType in_1) {
```
The vulkan interop tests also use these functions. can you remove them from the interop test and keep these in a single place? `add_kernel`, `fill_rand`, `printString`

isaacault replied:
Using these functions here but leaving `printString` as is for now as it's not included in this file.

cppchedy commented on sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:
```diff
@@ -0,0 +1,467 @@
+  // parallel_for 2D
+  template <int NDims, typename DType, int NChannels, typename KernelName,
+            typename = std::enable_if_t<NDims == 2>>
+  static void run_ndim_test(
```
I think `run_ndim_test` can be fused into a single function that handles both 2D and 3D cases. you can take a look at https://github.com/intel/llvm/blob/ae24111fc96146b7daa8aea6d09325ddee76c927/sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp#L144

cppchedy commented on sycl/include/sycl/ext/oneapi/bindless_images.hpp:
```diff
@@ -1030,6 +1030,45 @@ DataT read_image(const sampled_image_handle &imageHandle [[maybe_unused]],
+                       [[maybe_unused]],
+                       const CoordT &coords [[maybe_unused]],
+                       const int arrayLayer [[maybe_unused]]) {
+  constexpr size_t coordSize = detail::coord_size<CoordT>();
```
```suggestion
  detail::assert_unsampled_coords<CoordT>();
  constexpr size_t coordSize = detail::coord_size<CoordT>();
```

isaacault replied:
Good catch. Done.

isaacault commented on sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:
```diff
@@ -0,0 +1,467 @@
+
+// Uncomment to print additional test information
+// #define VERBOSE_PRINT
+
```
Good shout, done.

isaacault commented on sycl/test-e2e/bindless_images/bindless_helpers.hpp:
```diff
@@ -0,0 +1,59 @@
+namespace bindless_helpers {
+
+template <typename DType, int NChannels>
+static void fillRand(std::vector<sycl::vec<DType, NChannels>> &v, int seed) {
```
Changed.

jchlanda commented on libclc/ptx-nvidiacl/libspirv/images/image.cl:
```diff
@@ -256,36 +256,47 @@ pixelf32 as_pixelf32(int4 v) { return as_float4(v); }
-_DEFINE_VEC4_CAST(float, uint)
-_DEFINE_VEC4_CAST(uint, float)
-_DEFINE_VEC4_CAST(uint, int)
+_DEFINE_VEC4_CAST(int, int)
```
This doesn't look right, casting between `int4` and `int4'? Did you mean:
```cpp
_DEFINE_VEC4_CAST(uint, int)
```

jchlanda commented on libclc/ptx-nvidiacl/libspirv/images/image.cl:
```diff
@@ -256,36 +256,47 @@ pixelf32 as_pixelf32(int4 v) { return as_float4(v); }
+_DEFINE_VEC4_CAST(int, float)
 _DEFINE_VEC4_CAST(uint, ushort)
 _DEFINE_VEC4_CAST(uint, uchar)
+_DEFINE_VEC4_CAST(short, short)
```
Same here, `short4` to `short4`?

jchlanda commented on libclc/ptx-nvidiacl/libspirv/images/image.cl:
```diff
@@ -256,36 +256,47 @@ pixelf32 as_pixelf32(int4 v) { return as_float4(v); }
+_DEFINE_VEC4_TO_VEC2_CAST(int, uint)
 _DEFINE_VEC4_TO_VEC2_CAST(int, short)
 _DEFINE_VEC4_TO_VEC2_CAST(int, char)
+_DEFINE_VEC4_TO_VEC2_CAST(uint, uint)
```
Again, same type on both sides.

jchlanda commented on libclc/ptx-nvidiacl/libspirv/images/image.cl:
```diff
@@ -256,36 +256,47 @@ pixelf32 as_pixelf32(int4 v) { return as_float4(v); }
+_DEFINE_VEC4_TO_VEC2_CAST(uint, uint)
 _DEFINE_VEC4_TO_VEC2_CAST(uint, ushort)
 _DEFINE_VEC4_TO_VEC2_CAST(uint, uchar)
+_DEFINE_VEC4_TO_VEC2_CAST(short, short)
```
Again.

jchlanda commented on libclc/ptx-nvidiacl/libspirv/images/image.cl:
```diff
@@ -256,36 +256,47 @@ pixelf32 as_pixelf32(int4 v) { return as_float4(v); }
+_DEFINE_VEC4_TO_VEC2_CAST(short, short)
+_DEFINE_VEC4_TO_VEC2_CAST(short, ushort)
+_DEFINE_VEC4_TO_VEC2_CAST(short, char)
+_DEFINE_VEC4_TO_VEC2_CAST(float, float)
```
And again. A whole bunch more of the same type on both sides, going to ignore them from now on, but if they are indeed not needed, please address all the cases.

Seems that this patch adds some more casts, as well as reshuffles the existing ones. In an ideal git scenario, the reshuffling would be in a separate NFC PR, so it's easier to spot what is actually going on.

jchlanda replied:
Right, I understand now. Would you be bale to verify that the identity cast is indeed optimised out in the compiler and does not make its way to PTX?

jchlanda replied:
I guess it's your call, so I'm happy to go with your judgement, but it would appear to me that for the casts in question they are either:
* optimised out, so there is no problem,
* or are generating the cast instruction in which case the macro should be re-factored?

ldrumm replied:
This is the sort of work that never gets done, and there's no point refactoring if you have the opportunity to do it right the first time. If you can verify that the cast doesn't end up with extraneous code in the kernel now, I think that would placate Jakub somewhat. It might "just work" ;)

isaacault replied:
@jchlanda @ldrumm I understand your concerns.

I've looked into this and the output of `cuobjdump` called on my below e2e test is identical regardless of using these macros with the identity cast or an older version of `image.cl` where I don't have this redundancy.

With this, **I believe the identity cast is optimized out.**

Example PTX for the instance of reading 2 floats then writing their sum, where the results read would be cast with `cast_int_to_int()`.

```
suld.b.a1d.b32.clamp {%r10}, [%rd1, {%r8, %r9}];
mov.b32 %f1, %r10;                              
suld.b.a1d.b32.clamp {%r11}, [%rd2, {%r8, %r9}];
mov.b32 %f2, %r11;                              
add.rn.f32 %f3, %f1, %f2;                       
mov.b32 %r12, %f3;                              
sust.b.a1d.b32.clamp [%rd3, {%r8, %r9}], {%r12};
```

Note: This is my first time emitting and analysing PTX from one of the e2e tests, so for sanity check I removed the `as_float()` cast (which is used) and successfully observed this difference in the PTX emitted. That is to say, if you think my conclusion above isn't sufficient then please do share.

jchlanda replied:
Great, thank you for checking that, much appreciated.

jchlanda commented on libclc/ptx-nvidiacl/libspirv/images/image.cl:
```diff
@@ -2588,3 +2599,349 @@ _CLC_DEFINE_MIPMAP_BINDLESS_READS_BUILTIN(half4, 3, Dv4_DF16_, v4f16, Dv4_f, flo
+    "llvm.nvvm.suld.1d.array.i8.clamp");
+short __nvvm_suld_2d_array_i8_clamp_s_helper(long, int, int, int) __asm(
+    "llvm.nvvm.suld.2d.array.i8.clamp");
+short2 __nvvm_suld_1d_array_v2i8_clamp_s_helper(long, int, int) __asm(
```
With 4 cases of `__nvvm_suld_[ARRAY_D]_array_[TYPE]_clamp_s_helper` I'd be tempted to wrap it in a macro, such is libclc style anyway 😄

jchlanda commented on libclc/ptx-nvidiacl/libspirv/images/image.cl:
```diff
@@ -2588,3 +2599,349 @@ _CLC_DEFINE_MIPMAP_BINDLESS_READS_BUILTIN(half4, 3, Dv4_DF16_, v4f16, Dv4_f, flo
+#undef _CLC_DEFINE_SURFACE_ARRAY_BINDLESS_THUNK_READS_BUILTIN
+
+// //  --- THUNKS: Surface Array Writes ---
+// int
```
Same here, maybe we could wrap cases for `int`, `short` and `char`in their own macros, seems to be a common pattern in each of the cases.

jchlanda reviewed (commented):
I've only gone through libclc changes.

isaacault commented:
@jchlanda My apologies, I accidentally rebased. I believe I've corrected my mistake, and you can see my changes addressing your feedback in [6b2ca14](https://github.com/intel/llvm/pull/12464/commits/6b2ca14b2793cf606790296a46318e2ac3ab3516).

aelovikov-intel commented on sycl/source/detail/bindless_images.cpp:
```diff
@@ -28,11 +28,25 @@ void populate_pi_structs(const image_descriptor &desc, pi_image_desc &piDesc,
+    // Image Array
+    if (desc.depth > 0) {
+      // Image arrays must be 1D or 2D
+      throw sycl::exception(sycl::make_error_code(sycl::errc::invalid),
```
Is it missing support or they simply can't be per the specification? If the latter, why can't we rely on `verify` to throw instead? Same in `handler.cpp`.

aelovikov-intel commented on sycl/source/handler.cpp:
```diff
@@ -1002,10 +1017,24 @@ void handler::ext_oneapi_copy(
-                                                    : PI_MEM_TYPE_IMAGE1D);
+  PiDesc.image_array_size = DestImgDesc.array_size;
+
+  if (DestImgDesc.array_size > 1) {
+    // Image Array
+    if (DestImgDesc.depth > 0) {
+      // Image arrays must be 1D or 2D
+      throw sycl::exception(
+          sycl::make_error_code(sycl::errc::invalid),
+          "Image descriptor malformed - cannot copy 3D image arrays.");
+    }
+    PiDesc.image_type = DestImgDesc.height > 0 ? PI_MEM_TYPE_IMAGE2D_ARRAY
+                                               : PI_MEM_TYPE_IMAGE1D_ARRAY;
+  } else {
+    PiDesc.image_type = DestImgDesc.depth > 0
+                            ? PI_MEM_TYPE_IMAGE3D
+                            : (DestImgDesc.height > 0 ? PI_MEM_TYPE_IMAGE2D
+                                                      : PI_MEM_TYPE_IMAGE1D);
+  }
```
How is this different from line 969-896? Can we outline into a helper to avoid copy-paste/make distinctins (if any) clearer? Same in multiple places below.

aelovikov-intel commented on sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:
```diff
@@ -0,0 +1,457 @@
+// REQUIRES: linux
+// REQUIRES: cuda
```
1) How is this specific to linux? I don't think we have windows+cuda yet at all.
2) nit, it's possible to `REQUIRES: linux, cuda`

aelovikov-intel commented on sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:
```diff
@@ -0,0 +1,457 @@
+// REQUIRES: linux
+// REQUIRES: cuda
+
+// RUN: %clangxx -fsycl -fsycl-targets=%{sycl_triple} %s -o %t.out
```
```suggestion
// RUN: %{build} -o %t.out
```

aelovikov-intel commented on sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:
```diff
@@ -0,0 +1,457 @@
+// REQUIRES: cuda
+
+// RUN: %clangxx -fsycl -fsycl-targets=%{sycl_triple} %s -o %t.out
+// RUN: %t.out
```
```suggestion
// RUN: %{run} %t.out
```

aelovikov-intel commented on sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:
```diff
@@ -0,0 +1,457 @@
+
+static sycl::device dev;
+
+// Uncomment to print additional test information
+// #define VERBOSE_PRINT
```
```suggestion
#define VERBOSE_PRINT 0
```

and update the checks accordingly. Or (even better) switch to `inline constexpr bool` and `if constexpr`.

aelovikov-intel commented on sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:
```diff
@@ -0,0 +1,457 @@
+  // parallel_for 2D
+  template <int NDims, typename DType, int NChannels, typename KernelName,
+            typename = std::enable_if_t<NDims == 2>>
+  static void run_ndim_test(sycl::queue q, sycl::range<2> globalSize,
```
I think metaprogramming magic can unify these two versions neatly, but I won't insist.

aelovikov-intel commented on sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp:
```diff
@@ -0,0 +1,457 @@
+  sycl::queue q(dev);
+  auto ctxt = q.get_context();
+
+  // skip half tests if not supported
```
I'd add a `static_assert` against `double` that would otherwise require `aspect:fp64`.

jchlanda commented:
> @jchlanda My apologies, I accidentally rebased. I believe I've corrected my mistake, and you can see my changes addressing your feedback in [6b2ca14](https://github.com/intel/llvm/pull/12464/commits/6b2ca14b2793cf606790296a46318e2ac3ab3516).

Not a problem, I've left one more question, otherwise the libclc part looks good. Thank you.

aarongreig commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -56,14 +56,13 @@ endif()
-  #     Merge pull request #1249 from GeorgeWeb/georgi/hip_memadvise_coarse_grained
-  #     [HIP] Implement coarse-grained memory advice for the HIP adapter
-  set(UNIFIED_RUNTIME_TAG cd97e172cbbfc411fccb0b80e0fff6f9126574f4)
+  set(UNIFIED_RUNTIME_REPO "https://github.com/isaacault/unified-runtime.git")
+  # commit f6ab4ed9846a472d8cb173c12fb9affae21c0fd0
+  # Author: Isaac Ault <isaac.ault@codeplay.com>
+  # Date:   Tue Jan 23 12:34:08 2024 +0000
+  #
+  #     [Bindless][Exp] Add Support For Image Arrays
+  set(UNIFIED_RUNTIME_TAG f6ab4ed9846a472d8cb173c12fb9affae21c0fd0)
```
https://github.com/orgs/intel/teams/unified-runtime-reviewers will not approve until this is updated, but we do have a policy in our repo to get approvals from other reviewers here before contributions there can proceed. We take full responsibility for blocking this kind of change

aelovikov-intel reviewed (approved):
SYCL RT changes are good, cursory reading of e2e tests is fine stylistically, I'd expect the logic/completeness of the tests to be checked by images domain experts from Codeplay.

jchlanda reviewed (approved):
libclc 👍

kbenzie commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -56,14 +56,13 @@ endif()
-  #     Merge pull request #1299 from rafbiels/rafbiels/fix-cuda-maxreg-check
-  #     [CUDA] Fix MaxRegsPerBlock check in setKernelParams
-  set(UNIFIED_RUNTIME_TAG 24078c26ab871572067f520d856f4c65271cb9e5)
+  set(UNIFIED_RUNTIME_REPO "https://github.com/isaacault/unified-runtime.git")
+  # commit f6ab4ed9846a472d8cb173c12fb9affae21c0fd0
+  # Author: Isaac Ault <isaac.ault@codeplay.com>
+  # Date:   Tue Jan 23 12:34:08 2024 +0000
+  #
+  #     [Bindless][Exp] Add Support For Image Arrays
+  set(UNIFIED_RUNTIME_TAG 89ad3408f09ed22249f051795c91287f342193b4)
```
```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit 91c6068bb581acf72f02f1284ab0464ec4f03e34
  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
  # Date:   Fri Feb 23 13:39:03 2024 +0000
  #     Merge pull request #1274 from isaacault/iault/image_array_copy
  #     [Bindless][Exp] Add Support For Image Arrays
  set(UNIFIED_RUNTIME_TAG 91c6068bb581acf72f02f1284ab0464ec4f03e34)
```

kbenzie reviewed (commented):
https://github.com/oneapi-src/unified-runtime/pull/1274 has been merged, please pull in the latest sycl branch changes and update the UR tag as suggested.

isaacault commented:
@intel/llvm-gatekeepers Can this get merged? CI failure unrelated & already reported in https://github.com/intel/llvm/issues/12797 and https://github.com/intel/llvm/issues/12798

kbenzie commented:
@intel/llvm-gatekeepers please merge, the Windows failure is being tracked in #12798
